File: backend/GcePlatform.Api/Helpers/SubmissionValidator.cs
using System.Text.RegularExpressions;

namespace GcePlatform.Api.Helpers;

/// <summary>
/// Per-assignment validation: range, precision, regex. All rules optional.
/// Authoritative server-side check, mirrored client-side for instant UX.
/// usp_SubmitKpi only snapshots the rules; the validator runs in the endpoint.
/// </summary>
public static class SubmissionValidator
{
    public record Rules(
        decimal? MinValue,
        decimal? MaxValue,
        int?     Precision,
        string?  Regex,
        string?  Message);

    public record Result(bool Ok, string? ErrorMessage);

    public static readonly Result Pass = new(true, null);

    /// <summary>
    /// Validates a submission's payload against the assignment's rules.
    /// Returns <see cref="Pass"/> when no rule of the relevant kind is set,
    /// otherwise the first failing rule's message (custom if provided).
    /// </summary>
    public static Result Validate(string dataType, Rules rules,
                                  decimal? value, string? text)
    {
        // Numeric-like data types: range + precision on `value`.
        if (dataType is "Numeric" or "Percentage" or "Currency" or "Time")
        {
            if (value is null)
                return Pass; // Empty submission handled by IsRequired logic, not here.

            if (rules.MinValue is not null && value < rules.MinValue)
                return Fail(rules, $"Value must be at least {rules.MinValue}.");
            if (rules.MaxValue is not null && value > rules.MaxValue)
                return Fail(rules, $"Value must be at most {rules.MaxValue}.");
            if (rules.Precision is not null)
            {
                var rounded = Math.Round(value.Value, rules.Precision.Value);
                if (rounded != value.Value)
                {
                    var places = rules.Precision.Value;
                    return Fail(rules, places == 0
                        ? "Value must be a whole number."
                        : $"Value must have at most {places} decimal place{(places == 1 ? "" : "s")}.");
                }
            }
            return Pass;
        }

        // Text / DropDown — regex only.
        if (dataType is "Text" or "DropDown")
        {
            if (rules.Regex is null || string.IsNullOrEmpty(rules.Regex))
                return Pass;
            if (text is null) return Pass;

            try
            {
                var match = Regex.IsMatch(
                    text,
                    rules.Regex,
                    RegexOptions.None,
                    matchTimeout: TimeSpan.FromMilliseconds(100));
                if (!match)
                    return Fail(rules, "Value does not match the required format.");
            }
            catch (RegexMatchTimeoutException)
            {
                return Fail(rules, "Validation regex took too long to evaluate. Contact your admin.");
            }
            catch (ArgumentException)
            {
                // Malformed regex on the assignment — surface to the admin via the error.
                return Fail(rules, "Validation regex is invalid. Contact your admin.");
            }
            return Pass;
        }

        // Boolean / unknown types: nothing to validate.
        return Pass;
    }

    private static Result Fail(Rules rules, string defaultMessage)
        => new(false, string.IsNullOrWhiteSpace(rules.Message) ? defaultMessage : rules.Message);
}

File: frontend/src/lib/submission-validator.ts
/**
 * Mirror of backend `SubmissionValidator.cs` so the capture form can give
 * instant red-text feedback before the user submits. The backend is still the
 * authoritative check on POST /kpi/submissions.
 */

export interface ValidationRules {
  minValue: number | null
  maxValue: number | null
  precision: number | null
  regex: string | null
  message: string | null
}

export type ValidationResult = { ok: true } | { ok: false; error: string }

const PASS: ValidationResult = { ok: true }

export function validateSubmission(args: {
  dataType: string
  rules: ValidationRules
  value?: number | null
  text?: string | null
}): ValidationResult {
  const { dataType, rules, value, text } = args

  // Numeric-like: range + precision check on `value`.
  if (
    dataType === 'Numeric' ||
    dataType === 'Percentage' ||
    dataType === 'Currency' ||
    dataType === 'Time'
  ) {
    if (value === null || value === undefined) return PASS

    if (rules.minValue !== null && value < rules.minValue)
      return fail(rules, `Value must be at least ${rules.minValue}.`)
    if (rules.maxValue !== null && value > rules.maxValue)
      return fail(rules, `Value must be at most ${rules.maxValue}.`)
    if (rules.precision !== null) {
      // Match the C# Math.Round semantics: round half-to-even is JS default
      // for toFixed via Number, but for our purposes we just check whether
      // any digit beyond `precision` is non-zero.
      const factor = Math.pow(10, rules.precision)
      if (Math.round(value * factor) / factor !== value) {
        return fail(
          rules,
          rules.precision === 0
            ? 'Value must be a whole number.'
            : `Value must have at most ${rules.precision} decimal place${rules.precision === 1 ? '' : 's'}.`,
        )
      }
    }
    return PASS
  }

  // Text / DropDown — regex only.
  if (dataType === 'Text' || dataType === 'DropDown') {
    if (!rules.regex) return PASS
    if (text === null || text === undefined) return PASS
    try {
      const re = new RegExp(rules.regex)
      if (!re.test(text))
        return fail(rules, 'Value does not match the required format.')
    } catch {
      return fail(rules, 'Validation regex is invalid. Contact your admin.')
    }
    return PASS
  }

  // Boolean / unknown: nothing to validate.
  return PASS
}

function fail(rules: ValidationRules, defaultMessage: string): ValidationResult {
  const m = rules.message?.trim()
  return { ok: false, error: m && m.length > 0 ? m : defaultMessage }
}
